hpx/parallel/segmented_algorithms/reduce.hpp

```diff
@@ -0,0 +1,252 @@
+            segments.reserve(std::distance(sit, send));
+
+            if (sit == send)
+            {
+                // all elements are on the same partition
+                local_iterator_type beg = traits::local(first);
+                local_iterator_type end = traits::local(last);
+                if (beg != end)
+                {
+                    segments.push_back(
+                        dispatch_async(traits::get_id(sit),
+                            algo, policy, forced_seq(),
+                            beg, end, init, red_op)
+                    );
+                }
+            }
+            else {
+                // handle the remaining part of the first partition
+                local_iterator_type beg = traits::local(first);
+                local_iterator_type end = traits::end(sit);
+                if (beg != end)
+                {
+                    segments.push_back(
+                        dispatch_async(traits::get_id(sit),
+                            algo, policy, forced_seq(),
+                            beg, end, init, red_op)
+                    );
+                }
+
+                // handle all of the full partitions
+                for (++sit; sit != send; ++sit)
+                {
+                    beg = traits::begin(sit);
+                    end = traits::end(sit);
+                    if (beg != end)
+                    {
+                        segments.push_back(
+                            dispatch_async(traits::get_id(sit),
+                                algo, policy, forced_seq(),
+                                beg, end, init, red_op)
+                        );
+                    }
+                }
+
+                // handle the beginning of the last partition
+                beg = traits::begin(sit);
+                end = traits::local(last);
+                if (beg != end)
+                {
+                    segments.push_back(
+                        dispatch_async(traits::get_id(sit),
+                            algo, policy, forced_seq(),
+                            beg, end, init, red_op)
+                    );
+                }
+            }
+
+            return result::get(
+                dataflow(
+                    [=](std::vector<shared_future<T> > && r) -> T
+                    {
+                        // handle any remote exceptions, will throw on error
+                        std::list<std::exception_ptr> errors;
+                        parallel::util::detail::handle_remote_exceptions<
+                            ExPolicy
+                        >::call(r, errors);
+
+                        // VS2015RC bails out if red_op is capture by ref
+                        return std::accumulate(
+                            r.begin(), r.end(), init,
+                            [=](T const& val, shared_future<T>& curr)
+                            {
+                                return red_op(val, curr.get());
```

Comment:
I have already done this here: https://github.com/STEllAR-GROUP/hpx/blob/fixing_2699/hpx/parallel/algorithms/is_heap.hpp#L76.

The reason to use `invoke` that that it allows to transparently pass any callable to the algorithm, include member function pointers, in which case the current element is used as the `this` for the function invocation.